### PR TITLE
[CONTRACTS] Check side effect of loop contracts during instrumentation

### PR DESCRIPTION
**Pull request by qinheping**
Now, CBMC check side effect of loop contracts in `goto_convertt` by rejecting all `side_effect_exprt`. However, we might want to accept side-effect loop contracts and remove the side effect during instrumentation, or accept `side_effect_exprt` which is actually side-effect free (see the added regression test of this PR as an example).

This PR postpone the check of side effect of loop contracts from code converting to the actual instrumentation time. It also provide a new options to disable the side-effect check.

This PR include the commit from #8359. The regression test added in this PR also test #8359 (introducing new symbols) and #8282 (containing a GOTO jump in the loop invariant). 

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/contracts-dfcc/loop_contracts_statement_expression/test.desc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--dfcc main --apply-loop-contracts --disable-side-effect-check
```
Comment (remi-delmas-3000) [inline regression/contracts-dfcc/loop_contracts_statement_expression/test.desc]: naming the option like could be confusing to the user, the side effect checking is not disabled, it is should still happen at some point.

File: src/goto-instrument/contracts/contracts.h [hunk 2]
```diff
@@ -33,6 +33,10 @@ Date: February 2016
 #define HELP_LOOP_CONTRACTS                                                    \
   " {y--apply-loop-contracts} \t check and use loop contracts when provided\n"
 
+#define FLAG_DISABLE_SIDE_EFFECT_CHECK "disable-side-effect-check"
+#define HELP_DISABLE_SIDE_EFFECT_CHECK                                         \
+  " {y--disable-side-effect-check} \t UNSOUND OPTION. do not check \t "        \
```
Comment (remi-delmas-3000) [inline src/goto-instrument/contracts/contracts.h]: We should be able to instrument the GOTO program snippet that encodes the invariant evaluation using an empty write set and still check for side-effect freeness, as is done for contracts pre and post conditions.

File: src/goto-instrument/contracts/utils.cpp [hunk 3]
```diff
@@ -682,6 +682,48 @@ get_loop_head(const unsigned int target_loop_number, goto_functiont &function)
...
 
+exprt get_loop_invariants(
+  const goto_programt::const_targett &loop_end,
+  const bool check_side_effect)
+{
+  auto invariant = static_cast<const exprt &>(
+    loop_end->condition().find(ID_C_spec_loop_invariant));
+  if(!invariant.is_nil() && check_side_effect)
+  {
+    if(has_subexpr(invariant, ID_side_effect))
```
Comment (remi-delmas-3000) [inline src/goto-instrument/contracts/utils.cpp:713]: Can we separate the check for side-effect expressions from the accessor function ?

**Review by remi-delmas-3000 — CHANGES_REQUESTED** [review]
Allowing arbitrary `side_effect_exprt` allows things like `__CPROVER_is_fresh` and `__CPROVER_r_ok` to be used in the loop invariant. Since `__CPROVER_r_ok` might not work as expected in assumption contexts, and since the semantics if `__CPROVER_is_fresh` is only well defined in the context of requires and ensures clauses relative to some call site, we still may want to ban them for now.

Also if when using DFCC, all functions of the model are extended with a write set parameter, and any function called in the loop invariant must be passed either a write set pointer, either NULL (in which case no checks will be done) or a valid pointer to some write set (and checks will be performed).

But we  should try to actually check for side effects _semantically_ using DFCC and an empty write set.

We already do it for preconditions and postconditions in function contracts 
[see here]( https://github.com/diffblue/cbmc/blob/develop/src/goto-instrument/contracts/dynamic-frames/dfcc_wrapper_program.cpp#L259)

You can declare and build the write set in PREFIX program
```C
// PREFIX
DECL ws : write_set_t ;
DECL ws_ptr : write_set_t* ;
ASSIGN ws_ptr = &ws;
CALL write_set_create(
ws_ptr,
assigns_size = 0,
frees_size = 0,
replacement_mode = false, ... );
```

Then you goto-convert the invariant clause and instrument the resulting program snippet against the empty write set:

```C
 // will contain the invariant evaluation
goto_programt loop_inv_prog;
converter.goto_convert(loop_inv, loop_inv_prog, language_mode);
// fix calls to memory predicates like `__CPROVER_is_fresh`
memory_predicates.fix_calls(loop_inv_prog);
loop_inv_prog.add(goto_programt::make_end_function());
instrument.instrument_goto_program(
    prog_id,
    loop_inv_prog,
    ws_ptr, // symbol that represents the empty loop invariant write set
    function_pointer_contracts);
// turn it back into a sequence of statements
requires_program.instructions.back().turn_into_skip();
```

Then generate a SUFFIX program that checks that evaluating the invariant did not allocate of free any memory, release the write set :

```C
// SUFFIX
DECL __check_no_alloc: bool;
// checks that evaluating the invariant did not use malloc or free
CALL __check_no_alloc = check_empty_alloc_dealloc(requilres_write_set);
ASSERT __check_no_alloc;
DEAD __check_no_alloc: bool;
CALL write_set_release(ws_ptr); // releases resources used by the write set.
DEAD ws_ptr;
DEAD ws;
```

Finally concatenate the PREFIX, invariant evaluation program and SUFFIX, and insert the into the program at the base case and step case locations.

This will ensure that the invariant expression has no side effects, and since it uses the global DFCC instrumentation system, it will also work if the `side_effect_exprt` of the invariant is a call to a function that also calls other functions, etc.

**Comment by remi-delmas-3000** [comment]
Allowing arbitrary `side_effect_exprt` allows things like `__CPROVER_is_fresh` and `__CPROVER_r_ok` to be used in the loop invariant. Since `__CPROVER_r_ok` might not work as expected in assumption contexts, and since the semantics if `__CPROVER_is_fresh` is only well defined in the context of pre and post conditions relative to some call site, we still may want to ban them for now.

File: src/goto-instrument/contracts/utils.cpp (hunk 3 above)
Comment (qinheping) [reply src/goto-instrument/contracts/utils.cpp:713]: I think we should have only one public function to extract loop contracts and check side-effect. But yes we can separate the accessor  functions as static private functions.

File: regression/contracts-dfcc/loop_contracts_statement_expression/test.desc (hunk 1 above)
Comment (qinheping) [reply regression/contracts-dfcc/loop_contracts_statement_expression/test.desc]: The purpose of this flag is to disable side-effect checks now for the sake of experimenting other features, e.g., kani loop contracts and statement-expression loop contracts while we develop the new side-effect checks.

File: src/goto-instrument/contracts/contracts.h (hunk 2 above)
Comment (qinheping) [reply src/goto-instrument/contracts/contracts.h]: As the comment above, this flag allow us to disable side-effect checks while we develop the new checks.

**Comment by qinheping** [comment]
Thank you @remi-delmas-3000 ! Yes, I think it is really good idea to apply the dfcc side-effect checks here.
However, it should be implemented in a following PR. This PR only moves the current checks from typechecking time to instrumentation time, and add a flag to disable side-effect check to allow us experiment kani loop contracts while we develop the new side-effect checks.

File: doc/man/goto-instrument.1 [hunk 4]
```diff
@@ -1015,7 +1015,9 @@ force aggressive slicer to preserve all direct paths
 .SS "Code contracts:"
 .TP
 \fB\-\-apply\-loop\-contracts\fR
-check and use loop contracts when provided
+.TP
+\fB\-disable\-side\-effect\-check\fR
+UNSOUND OPTION. do not check the side-effect of loop contracts
```
Comment (remi-delmas-3000) [inline doc/man/goto-instrument.1]: Suggestion: 

"UNSOUND OPTION. Disables checking the absence of side effects in loop contract clauses."

File: src/goto-instrument/contracts/contracts.h (hunk 2 above)
Comment (remi-delmas-3000) [inline src/goto-instrument/contracts/contracts.h]: suggestion `--disable-loop-contract-side-effect-check`

File: doc/man/goto-instrument.1 [hunk 5]
```diff
@@ -1015,7 +1015,9 @@ force aggressive slicer to preserve all direct paths
 .SS "Code contracts:"
 .TP
 \fB\-\-apply\-loop\-contracts\fR
-check and use loop contracts when provided
+.TP
+\fB\-disable\-side\-effect\-check\fR
```
Comment (remi-delmas-3000) [inline doc/man/goto-instrument.1]: Suggestion `--disable-loop-contract-side-effect-check` makes it more obvious that it is for loop contracts only, not function contracts.

**Review by remi-delmas-3000 — APPROVED** [review]
One minor optional comment about option naming.

File: doc/man/goto-instrument.1 (hunk 4 above)
Comment (tautschnig) [reply doc/man/goto-instrument.1]: Please add an explanation as to why anyone would want to use this option.

File: doc/man/goto-instrument.1 (hunk 5 above)
Comment (tautschnig) [reply doc/man/goto-instrument.1]: Elsewhere, the context may be even less obvious, so +100 to what Remi said.

File: regression/contracts-dfcc/loop_contracts_statement_expression/main.c [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#include <stdlib.h>
```
Comment (tautschnig) [inline regression/contracts-dfcc/loop_contracts_statement_expression/main.c]: Why is this include needed?

File: regression/contracts/loop_contracts_statement_expression/main.c [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#include <stdlib.h>
```
Comment (tautschnig) [inline regression/contracts/loop_contracts_statement_expression/main.c]: Why is this needed?

**Review by tautschnig — APPROVED** [review]
I'm ok with this other than as indicated in the comment below, provided that there will be follow-up work that will make this option completely unnecessary and we instead check for which side effects we genuinely cannot tolerate.
